README
Notes about the GtkDrop widget

Copyright (C) 2008 Gaute Hope <eg@gaute.vetsj.com>

GtkDrop is a widget I created when I wanted to create a popup calendar controlled by a toggle button, and as the example program shows it is ideal for this. You can use the GtkDrop widget as any other container as it derives from GtkWindow. And when you have set some options for how to calculate its position (see 'Positioning' further down) the widget will take care of most things itself.

When you click anywhere on the parentwindow the widget will be hidden, and when you move the parentwindow the position will be recalculated. You could easily change this behaviour to hiding the widget on moved parent window by editing 'gtkdrop.c' and the function 'gtk_drop_parent_moved' and change the line 'gtk_drop_refresh_position' with 'gtk_drop_hide'. If this widget ever gets interesting I plan to make this functionality automatic through a configuration setting.

Tip:
It might look nicer if you place all your widget inside a GtkFrame.

Building:
I set up a scons build environment for the test program, just run 'scons' in the source directory.

Usage:
Include 'gtkdrop.h', see 'test.c' for an example.

GtkDrop is using the normal gtk+ libraries, you get all needed compile arguments as normal by running: 'pkg-config --cflags --libs gtk+-2.0'.

Design notes:

Initializing:
Set up your window, in this example 'main_window'.
Create the widget:

  /* GtkDrop relies on the parent_window being able to move or close on events from this window */
  
  GtkWidget * drop = gtk_drop_new (main_window);
  
  /* This is set to GTK_DROP_ALIGN_DOWN by default and decides on what side of the position widget the widget will appear */
  
  gtk_drop_set_align (drop, GTK_DROP_ALIGN_RIGHT);
  
  /* This widget is used to calculate where the widget will appear, and it will be on one of the sides depending on the specified align */
  
  gtk_drop_set_position_widget (drop, for_instance_an_toggle_button);
  
  /* This is the margin from the position widget to the GtkDrop widget */
  
  gtk_drop_set_margin (drop, 5);


Signals:

GtkDrop emits the following signals:
  'drop-shown' - when the widget is shown
  'drop-hidden' - when the widget is hidden


Positioning:
The positioning of the GtkDrop widget can be calculated automatically by setting a 'position widget'. This is done with the function 'gtk_drop_set_position_widget', if the widget is allready shown you need to manually call 'gtk_drop_refresh_position' to recalculate the position (unless some event triggers it). This because in many cases the 'position widget' will not yet have been shown and does not yet have a GdkWindow associated with it.

You can manually calculate the position of the widget by assigning your own calculate function to GtkDrop with the 'gtk_drop_set_position_function'. 

The functions '*position_function' and '*position_widget' disables each other.

- gaute



ChangeLog
2008-02-04  Gaute Hope  <eg@gaute.vetsj.com>

	* long time no see, something working (useful ?) is ready.
	* wrote a README

2007-12-03  Gaute Hope  <eg@gaute.vetsj.com>

	* Set up intial everything, added gplv2 as license.
	* Created initial repository, build and file layout


test.c
/*
 * test.c - Test window for the GtkDrop widget
 *
 * Copyright (C) 2008 Gaute Hope <eg@gaute.vetsj.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.

 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 */

# include <stdlib.h>
# include <stdio.h>
# include <glib.h>
# include <gtk/gtk.h>
# include <gdk/gdkkeysyms.h>

# include "gtkdrop.h"

typedef struct _TestDrop TestDrop;

struct _TestDrop {
	GtkWindow *win;
	GtkDrop *drop;
	GtkWidget *button;
};

gboolean quit (GtkWidget *w, GdkEvent *e, gpointer *data);
gboolean key_pressed (GtkWidget *w, GdkEventKey *e, gpointer *data);
gboolean button_toggled (GtkWidget *w, GdkEvent *e, gpointer *data);
gboolean drop_hidden (GtkWidget *w, GdkEvent *e, gpointer *data);
void calc_drop_pos (int *x, int *y);

static TestDrop *test = NULL;

int main (int argc, char ** argv)
{
  printf ("Test program for the GtkDrop widget\n");
	printf ("Copyright (C) 2008 Gaute Hope <eg@gaute.vetsj.com>\n");

  gtk_init (&argc, &argv);
  	
  test = (TestDrop*) malloc (sizeof (TestDrop));

  GtkWidget *w = GTK_WIDGET(gtk_window_new (GTK_WINDOW_TOPLEVEL));
  test->win = GTK_WINDOW(w);
  	
  gtk_widget_show (GTK_WIDGET(test->win));
  gtk_window_set_resizable (GTK_WINDOW (test->win), FALSE);
  gtk_window_set_title (GTK_WINDOW (test->win), "GtkDrop test");

  g_signal_connect (G_OBJECT (test->win), "delete-event", G_CALLBACK (quit), NULL);
  g_signal_connect (G_OBJECT (test->win), "key-press-event", G_CALLBACK (key_pressed), NULL);

  GtkWidget *button = gtk_toggle_button_new ();
  g_signal_connect (G_OBJECT (button), "toggled", G_CALLBACK (button_toggled), NULL);
  test->button = button;

	GtkWidget *button_label = gtk_label_new ("Toggle");	

  GtkWidget *hbox = gtk_hbox_new (FALSE, 5);
  gtk_box_pack_start (GTK_BOX (hbox), button_label, FALSE, FALSE, 5);
  GtkWidget *arrow = gtk_arrow_new (GTK_ARROW_UP, GTK_SHADOW_IN);
  gtk_box_pack_start (GTK_BOX (hbox), arrow, FALSE, FALSE, 5);

  GtkWidget *mvbox = gtk_vbox_new (FALSE, 50);
  GtkWidget *mhbox = gtk_hbox_new (FALSE, 50);

  gtk_container_add (GTK_CONTAINER (button), hbox);
  gtk_box_pack_start (GTK_BOX (mhbox), button, FALSE, FALSE, 50);
  gtk_box_pack_start (GTK_BOX (mvbox), mhbox, FALSE, FALSE, 50);
 	gtk_container_add (GTK_CONTAINER (w), mvbox);

 	GtkWidget *drop = gtk_drop_new (GTK_WINDOW(w));
 	test->drop = GTK_DROP (drop);
 	gtk_drop_set_position_widget (test->drop, button);
 	gtk_drop_set_align (test->drop, GTK_DROP_ALIGN_UP);
 	
 	g_signal_connect (G_OBJECT (drop), "drop-hidden", G_CALLBACK (drop_hidden), NULL);
  	  	
 	GtkFrame *frame = GTK_FRAME (gtk_frame_new (NULL));
 	gtk_container_add (GTK_CONTAINER (drop), GTK_WIDGET (frame));
 	gtk_widget_show (GTK_WIDGET (frame));
  	
  GtkWidget *e_hbox = gtk_hbox_new (FALSE, 5);
  
 	GtkWidget *l = gtk_label_new ("Test label:");
 	gtk_box_pack_start (GTK_BOX (e_hbox), l, FALSE, FALSE, 5);
 	
 	GtkWidget *e = gtk_entry_new ();
 	gtk_box_pack_start (GTK_BOX (e_hbox), e, FALSE, FALSE, 5);
 	
 	GtkWidget *b = gtk_button_new_with_label ("Test");
 	gtk_box_pack_start (GTK_BOX (e_hbox), b, FALSE, FALSE, 5);
 	
 	
 	gtk_container_add (GTK_CONTAINER (frame), e_hbox);
 	
 	gtk_widget_show_all (e_hbox);
  	
 	gtk_widget_show_all (w);

 	gtk_main ();
 	return 0;
}

void calc_drop_pos (int *x, int *y)
{
	
}

gboolean drop_hidden (GtkWidget *w, GdkEvent *e, gpointer *data)
{
  gtk_toggle_button_set_active (GTK_TOGGLE_BUTTON (test->button), FALSE);
}

gboolean button_toggled (GtkWidget *w, GdkEvent *e, gpointer *data)
{
	if (gtk_toggle_button_get_active (GTK_TOGGLE_BUTTON (w))) {
		gtk_drop_show (test->drop);
	} else {
		gtk_drop_hide (test->drop);
	}
}

gboolean key_pressed (GtkWidget *w, GdkEventKey *e, gpointer *data)
{
  if (e->type == GDK_KEY_PRESS) {
    if (e->keyval == GDK_Escape) {
      gtk_main_quit ();
    }
  }
  return TRUE;
}

gboolean quit (GtkWidget *w, GdkEvent *e, gpointer *data)
{
  gtk_main_quit ();
  return TRUE;
}



gtkdrop.h
/*
 * gtkdrop.h - Header for the GtkDrop widget
 *
 * Copyright (C) 2008 Gaute Hope <eg@gaute.vetsj.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.

 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 */

# ifndef _GTK_DROP_H_
# define _GTK_DROP_H_

# include <gdk/gdk.h>
# include <gtk/gtkwindow.h>


# define GTK_TYPE_DROP 			  (gtk_drop_get_type ())
# define GTK_DROP(obj)          GTK_CHECK_CAST (obj, gtk_drop_get_type (), GtkDrop)
# define GTK_DROP_CLASS(klass)  GTK_CHECK_CLASS_CAST (klass, gtk_drop_get_type (), GtkDropClass)
# define GTK_IS_DROP(obj)       GTK_CHECK_TYPE (obj, gtk_drop_get_type ())


typedef struct _GtkDrop       GtkDrop;
typedef struct _GtkDropClass  GtkDropClass;

typedef enum {
	GTK_DROP_ALIGN_UP,
	GTK_DROP_ALIGN_DOWN,
	GTK_DROP_ALIGN_RIGHT,
	GTK_DROP_ALIGN_LEFT
	
} GtkDropAlign;

struct _GtkDrop
{
	GtkWindow window;
	GtkWindow *parent_window;
	GtkWidget *pos_widget;
	GtkDropAlign align;

  int hidden;
  int auto_calc_pos;
  guint margin;

	GtkAllocation *allocation;
	GtkAllocation *parentalloc;

	
	void (*calculate_position)	(GtkDrop*, gint *, gint *);
};

struct _GtkDropClass
{
  GtkWindowClass parent_class;
  
  void (* gtkdrop) (GtkDrop *drop);
};

guint       gtk_drop_get_type        		  (void);
GtkWidget*  gtk_drop_new             		  (GtkWindow *);
void	      gtk_drop_show           		  (GtkDrop *drop);
void				gtk_drop_hide						      (GtkDrop *drop);
void				gtk_drop_set_position_widget 	(GtkDrop *drop, GtkWidget *widget);
void 				gtk_drop_set_position_function (GtkDrop *drop, void (*calc_func));
void				gtk_drop_refresh_position		  (GtkDrop *drop);
void				gtk_drop_set_align				    (GtkDrop *drop, GtkDropAlign align);

 
# endif



gtkdrop.c
/*
 * gtkdrop.c - GtkDrop widget
 *
 * Copyright (C) 2008 Gaute Hope <eg@gaute.vetsj.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.

 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 */

# include <stdlib.h>
# include <stdio.h>
# include <gtk/gtk.h>
# include <gdk/gdkkeysyms.h>

# include "gtkdrop.h"

static void gtk_drop_class_init 	(GtkDropClass *klass);
static void gtk_drop_init 			(GtkDrop *drop);

void gtk_drop_calculate_position (GtkDrop *drop);
gint gtk_drop_delete (GtkDrop *drop, GdkEvent *e, gpointer *data);
void gtk_drop_parent_event (GtkWidget *parent, GdkEvent *e, gpointer *data);
void gtk_drop_parent_moved (GtkWidget *parent, GdkEvent *e, gpointer *data);


static GtkWindowClass *parent_class = NULL;

enum {
  DROP_SHOWN_SIGNAL,
  DROP_HIDDEN_SIGNAL,
  LAST_SIGNAL
};

static guint drop_signals[LAST_SIGNAL] = { 0 };

static void gtk_drop_class_init (GtkDropClass *klass)
{
	// init any fields and set up signals
	GtkObjectClass *object_class;
	GtkWidgetClass *widget_class;
	GtkContainerClass *container_class;
	GtkBinClass *bin_class;
	
	object_class = (GtkObjectClass*) klass;
	widget_class = (GtkWidgetClass*) klass;
	container_class = (GtkContainerClass*) klass;
	bin_class = (GtkBinClass*) klass;
	
	parent_class = gtk_type_class (gtk_window_get_type ());	
	
	drop_signals[DROP_SHOWN_SIGNAL] = 
	  g_signal_new ("drop-shown",
	                G_TYPE_FROM_CLASS (klass),
	                G_SIGNAL_RUN_LAST | G_SIGNAL_ACTION,
	                G_STRUCT_OFFSET (GtkDropClass, gtkdrop),
	                NULL, NULL,
	                g_cclosure_marshal_VOID__VOID, G_TYPE_NONE, 0);
	                
	drop_signals[DROP_HIDDEN_SIGNAL] = 
	  g_signal_new ("drop-hidden",
	                G_TYPE_FROM_CLASS (klass),
	                G_SIGNAL_RUN_LAST | G_SIGNAL_ACTION,
	                G_STRUCT_OFFSET (GtkDropClass, gtkdrop),
	                NULL, NULL,
	                g_cclosure_marshal_VOID__VOID, G_TYPE_NONE, 0);
	                
}

static void gtk_drop_init (GtkDrop *drop)
{
	// set fields to standard & create widgets
	drop->parent_window = NULL;
	drop->hidden = TRUE;
	drop->align = GTK_DROP_ALIGN_DOWN;
	drop->auto_calc_pos = TRUE;
	drop->margin = 0;
	drop->allocation = malloc (sizeof (GtkAllocation));
	drop->allocation->x = 0;
	drop->allocation->y = 0;
	drop->allocation->width = 0;
	drop->allocation->height = 0;
	drop->pos_widget = NULL;
	drop->calculate_position = NULL;
	drop->parentalloc = malloc (sizeof (GtkAllocation));

	
	gtk_window_set_skip_taskbar_hint (GTK_WINDOW (drop), TRUE);
	gtk_window_set_skip_pager_hint (GTK_WINDOW (drop), TRUE);
	gtk_window_set_modal (GTK_WINDOW (drop), FALSE);
	gtk_window_set_decorated (GTK_WINDOW (drop), FALSE);
	gtk_window_set_deletable (GTK_WINDOW (drop), FALSE);
	gtk_window_set_type_hint (GTK_WINDOW (drop), GDK_WINDOW_TYPE_HINT_NORMAL);
	gtk_window_set_resizable (GTK_WINDOW (drop), FALSE);
	gtk_window_set_accept_focus (GTK_WINDOW (drop), TRUE);
	
	g_signal_connect (G_OBJECT (drop), "delete-event", G_CALLBACK (gtk_drop_delete), NULL);
  /* the rest of the signals will be set in gtk_drop_new, since they need the parent_window */
  
}

guint gtk_drop_get_type ()
{
  static guint drop_type = 0;

  if (!drop_type)
    {
      GtkTypeInfo drop_info =
      {
			"GtkDrop",
			sizeof (GtkDrop),
			sizeof (GtkDropClass),
			(GtkClassInitFunc) gtk_drop_class_init,
			(GtkObjectInitFunc) gtk_drop_init,
			 NULL,
      };

      drop_type = gtk_type_unique (GTK_TYPE_WINDOW, &drop_info);
    }

  return drop_type;
}


GtkWidget * gtk_drop_new (GtkWindow *parent_window)
{
	if (GTK_IS_WINDOW (parent_window)) {
		GtkDrop *drop = gtk_type_new (gtk_drop_get_type ());
		drop->parent_window = parent_window;
		gtk_window_set_transient_for (GTK_WINDOW (drop), parent_window);
		
	  gtk_widget_add_events (GTK_WIDGET (drop->parent_window), GDK_BUTTON_PRESS_MASK | GDK_KEY_PRESS_MASK | GDK_VISIBILITY_NOTIFY_MASK | GDK_STRUCTURE_MASK);
	  
		g_signal_connect (G_OBJECT (parent_window), "button-press-event", G_CALLBACK (gtk_drop_parent_event), drop);
		g_signal_connect (G_OBJECT (parent_window), "key-press-event", G_CALLBACK (gtk_drop_parent_event), drop);

		g_signal_connect (G_OBJECT (parent_window), "event-after", G_CALLBACK (gtk_drop_parent_moved), drop);

		
		g_signal_connect (G_OBJECT (parent_window), "delete-event", G_CALLBACK (gtk_drop_delete), drop);
		
		return GTK_WIDGET (drop);
	} else {
		return NULL;
	}
}

void gtk_drop_parent_moved (GtkWidget *parent, GdkEvent *e, gpointer *data)
{
  GtkDrop *drop = GTK_DROP (data);
  if (drop->hidden == FALSE) {
    gint x, y, height, width;
    gtk_window_get_position (GTK_WINDOW (drop->parent_window), &x, &y);
    gtk_window_get_size (GTK_WINDOW (drop->parent_window), &width, &height);
    if ((x != drop->parentalloc->x) || (y != drop->parentalloc->y) || (height != drop->parentalloc->height) || (width != drop->parentalloc->width)) 
    {
      //gtk_drop_hide (drop);
      gtk_drop_refresh_position (drop);
    }   
  }
}

void gtk_drop_parent_event (GtkWidget *parent, GdkEvent *e, gpointer *data)
{ 
  GtkDrop *drop = GTK_DROP (data);

  if (drop->hidden == FALSE) {
    gtk_drop_hide (drop);
  }
}

gint gtk_drop_delete (GtkDrop *drop, GdkEvent *e, gpointer *data)
{
  gtk_drop_hide (drop);
  return 1;
}

void gtk_drop_refresh_position (GtkDrop *drop)
{
  gtk_drop_calculate_position (drop);
  gtk_window_move (GTK_WINDOW (drop), drop->allocation->x, drop->allocation->y);
}

void gtk_drop_set_position_widget (GtkDrop *drop, GtkWidget *widget)
{
	drop->auto_calc_pos = TRUE;
	drop->calculate_position = NULL;
	drop->pos_widget = widget;
}

void gtk_drop_set_position_function (GtkDrop *drop, void *calc_func)
{
	drop->calculate_position = calc_func;
	drop->auto_calc_pos = FALSE;
}

void gtk_drop_show (GtkDrop *drop)
{
  if (drop->hidden) {
	  gtk_drop_refresh_position (drop);
	  gtk_widget_show (GTK_WIDGET (drop));
	  drop->hidden = FALSE;
	  g_signal_emit (G_OBJECT (drop),
	                 drop_signals[DROP_SHOWN_SIGNAL], 0);	
  }
}

void gtk_drop_hide (GtkDrop *drop) 
{
  if (drop->hidden == FALSE) {
	  gtk_widget_hide (GTK_WIDGET (drop));
	  if (GTK_IS_WINDOW (drop->parent_window)) {
  	}
  	g_signal_emit (G_OBJECT (drop),
	                 drop_signals[DROP_HIDDEN_SIGNAL], 0);	
	  drop->hidden = TRUE;
  }
}

void gtk_drop_set_align (GtkDrop *drop, GtkDropAlign align)
{
	drop->align = align;
}

void gtk_drop_calculate_position (GtkDrop *drop)
{
  gtk_window_get_position (GTK_WINDOW (drop->parent_window), &(drop->parentalloc->x), &(drop->parentalloc->y));
	gtk_window_get_size (GTK_WINDOW (drop->parent_window), &(drop->parentalloc->width), &(drop->parentalloc->height));
	  
	if ((drop->auto_calc_pos == TRUE) || (drop->calculate_position == NULL)) {
		if (drop->pos_widget) {
			gint x, y;
			GdkWindow *parent_window = gtk_widget_get_parent_window (drop->pos_widget);
			gdk_window_get_origin (parent_window, &x, &y);
			GtkAllocation *widget_alloc = &(drop->pos_widget->allocation);
				
			gtk_window_get_size (GTK_WINDOW (drop), &(drop->allocation->width), &(drop->allocation->height));
			
			switch (drop->align) {
				case GTK_DROP_ALIGN_UP:
					x = x + widget_alloc->x;
					y = y + widget_alloc->y - drop->allocation->height - drop->margin;
					break;
				default:
				case GTK_DROP_ALIGN_DOWN:
					x = x + widget_alloc->x;
					y = y + widget_alloc->y + widget_alloc->height + drop->margin;
					break;
				case GTK_DROP_ALIGN_RIGHT:
					x = x + widget_alloc->x + widget_alloc->width + drop->margin;
					y = y + widget_alloc->y;
					break;
				case GTK_DROP_ALIGN_LEFT:
					x = x + widget_alloc->x - drop->allocation->width - drop->margin;
					y = y + widget_alloc->y;
					break;
			}
			
			drop->allocation->x = x;
			drop->allocation->y = y;
			
		} else {
			drop->allocation->x = 0;
			drop->allocation->y = 0;
		}		
	} else {
		drop->calculate_position (drop, &(drop->allocation->x), &(drop->allocation->y));
	}
}

